### PR TITLE
[Synapse] - Build track 2 data plane SDK

### PR DESCRIPTION
**Pull request by idear1203**
This change aims to add the following 3 client libraries for Synapse:
- azure-analytics-synapse-spark
- azure-analytics-synapse-accesscontrol
- azure-analytics-synapse-artifacts

Most code is generated based on track 2 code generator. While we are still working on board review, we would like to build and release SDK packages in order to unblock our customers.

### CODE REVIEW

File: sdk/synapse/azure-analytics-synapse-accesscontrol/pom.xml [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  <parent>
+    <groupId>com.azure</groupId>
+    <artifactId>azure-client-sdk-parent</artifactId>
+    <version>1.7.0</version> <!-- {x-version-update;com.azure:azure-client-sdk-parent;current} -->
+    <relativePath>../../parents/azure-client-sdk-parent</relativePath>
+  </parent>
+
+  <groupId>com.azure</groupId>
+  <artifactId>azure-analytics-synapse-accesscontrol</artifactId>
+  <version>1.0.0-beta</version> <!-- {x-version-update;com.azure:azure-ai-formrecognizer;current} -->
```
Comment (mitchdenny) [inline sdk/synapse/azure-analytics-synapse-accesscontrol/pom.xml:16]: This should be ```1.0.0-beta.1```.

File: sdk/synapse/azure-analytics-synapse-development/pom.xml [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  <parent>
+    <groupId>com.azure</groupId>
+    <artifactId>azure-client-sdk-parent</artifactId>
+    <version>1.7.0</version> <!-- {x-version-update;com.azure:azure-client-sdk-parent;current} -->
+    <relativePath>../../parents/azure-client-sdk-parent</relativePath>
+  </parent>
+
+  <groupId>com.azure</groupId>
+  <artifactId>azure-analytics-synapse-development</artifactId>
+  <version>1.0.0-beta</version> <!-- {x-version-update;com.azure:azure-ai-formrecognizer;current} -->
```
Comment (mitchdenny) [inline sdk/synapse/azure-analytics-synapse-development/pom.xml:16]: This should be ```1.0.0-beta.1```.

File: sdk/synapse/pom.xml [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+     Licensed under the MIT License. -->
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+  <groupId>com.azure</groupId>
+  <artifactId>azure-synapse-service</artifactId>
+  <packaging>pom</packaging>
+  <version>1.0.0</version>  <!-- Need not change for every release-->
+  <modules>
+    <module>azure-analytics-synapse</module>
```
Comment (mitchdenny) [inline sdk/synapse/pom.xml:11]: Are you missing other modules here?

File: sdk/synapse/azure-analytics-synapse-accesscontrol/pom.xml (hunk 1 above)
Comment (idear1203) [reply sdk/synapse/azure-analytics-synapse-accesscontrol/pom.xml:16]: Fixed.

File: sdk/synapse/azure-analytics-synapse-development/pom.xml (hunk 2 above)
Comment (idear1203) [reply sdk/synapse/azure-analytics-synapse-development/pom.xml:16]: Fixed

File: sdk/synapse/pom.xml (hunk 3 above)
Comment (idear1203) [reply sdk/synapse/pom.xml:11]: Corrected the modules.

File: sdk/synapse/azure-analytics-synapse-accesscontrol/pom.xml [hunk 4]
```diff
@@ -13,7 +13,7 @@
 
   <groupId>com.azure</groupId>
   <artifactId>azure-analytics-synapse-accesscontrol</artifactId>
-  <version>1.0.0-beta</version> <!-- {x-version-update;com.azure:azure-ai-formrecognizer;current} -->
+  <version>1.0.0-beta.1</version> <!-- {x-version-update;com.azure:azure-ai-formrecognizer;current} -->
```
Comment (mitchdenny) [inline sdk/synapse/azure-analytics-synapse-accesscontrol/pom.xml:16]: Cut and paste error?

File: sdk/synapse/azure-analytics-synapse-development/pom.xml [hunk 5]
```diff
@@ -13,7 +13,7 @@
 
   <groupId>com.azure</groupId>
   <artifactId>azure-analytics-synapse-development</artifactId>
-  <version>1.0.0-beta</version> <!-- {x-version-update;com.azure:azure-ai-formrecognizer;current} -->
+  <version>1.0.0-beta.1</version> <!-- {x-version-update;com.azure:azure-ai-formrecognizer;current} -->
```
Comment (mitchdenny) [inline sdk/synapse/azure-analytics-synapse-development/pom.xml:16]: Cut and paste error?

File: sdk/synapse/azure-analytics-synapse-spark/pom.xml [hunk 6]
```diff
@@ -13,7 +13,7 @@
 
   <groupId>com.azure</groupId>
   <artifactId>azure-analytics-synapse-spark</artifactId>
-  <version>1.0.0-beta</version> <!-- {x-version-update;com.azure:azure-ai-formrecognizer;current} -->
+  <version>1.0.0-beta.1</version> <!-- {x-version-update;com.azure:azure-ai-formrecognizer;current} -->
```
Comment (mitchdenny) [inline sdk/synapse/azure-analytics-synapse-spark/pom.xml:16]: Cut and paste error?

**Comment by mitchdenny** [comment]
@JonathanGiles I know we've been talking about package name standards. Can I get your eyes on this to make the Java team are collectively happy with these new package names?

**Comment by JonathanGiles** [comment]
In terms of group names - in this case 'analytics' - we will need to raise it with the rest of the arch board to discuss. At the same time we should discuss the remainder of the naming pattern.

**Comment by idear1203** [comment]
> In terms of group names - in this case 'analytics' - we will need to raise it with the rest of the arch board to discuss. At the same time we should discuss the remainder of the naming pattern.

@JonathanGiles Sounds good to me. The "analytics" word comes from C# package naming rule. For your reference: https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-namespace-naming

File: sdk/synapse/azure-analytics-synapse-accesscontrol/pom.xml (hunk 4 above)
Comment (idear1203) [reply sdk/synapse/azure-analytics-synapse-accesscontrol/pom.xml]: Good catch! Fixed

File: sdk/synapse/azure-analytics-synapse-development/pom.xml (hunk 5 above)
Comment (idear1203) [reply sdk/synapse/azure-analytics-synapse-development/pom.xml]: Good catch! Fixed

File: sdk/synapse/azure-analytics-synapse-spark/pom.xml (hunk 6 above)
Comment (idear1203) [reply sdk/synapse/azure-analytics-synapse-spark/pom.xml]: Good catch! Fixed

**Comment by JonathanGiles** [comment]
> > In terms of group names - in this case 'analytics' - we will need to raise it with the rest of the arch board to discuss. At the same time we should discuss the remainder of the naming pattern.
> 
> @JonathanGiles Sounds good to me. The "analytics" word comes from C# package naming rule. For your reference: https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-namespace-naming

Got it! Analytics is also an approved group in Java too.

**Comment by idear1203** [comment]
Thank you @JonathanGiles ! May I also know who can help review the PR?

File: sdk/synapse/azure-analytics-synapse-accesscontrol/src/main/java/com/azure/analytics/synapse/accesscontrol/AccessControlAsyncClient.java [hunk 7]
```diff
@@ -0,0 +1,497 @@
...
+     * 
+     * @param roleAssignmentId The ID of the role assignment.
+     * @param context The context to associate with this operation.
+     * @throws IllegalArgumentException thrown if parameters fail the validation.
+     * @throws ErrorContractException thrown if the request is rejected by server.
+     * @throws RuntimeException all other wrapped checked exceptions if the request fails to be sent.
+     * @return role assignment by role assignment Id.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<SimpleResponse<RoleAssignmentDetails>> getRoleAssignmentByIdWithResponse(String roleAssignmentId, Context context) {
```
Comment (JonathanGiles) [inline sdk/synapse/azure-analytics-synapse-accesscontrol/src/main/java/com/azure/analytics/synapse/accesscontrol/AccessControlAsyncClient.java:216]: These should return `Response`, not `SimpleResponse`. Autorest was recently updated to return the correct type, so please update this API before GA.

File: sdk/synapse/azure-analytics-synapse-accesscontrol/src/main/java/com/azure/analytics/synapse/accesscontrol/AccessControlAsyncClient.java [hunk 8]
```diff
@@ -0,0 +1,497 @@
...
+     * List role assignments of the caller.
+     * 
+     * @param context The context to associate with this operation.
+     * @throws IllegalArgumentException thrown if parameters fail the validation.
+     * @throws ErrorContractException thrown if the request is rejected by server.
+     * @throws RuntimeException all other wrapped checked exceptions if the request fails to be sent.
+     * @return array of Post200ApplicationJsonItemsItem.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<SimpleResponse<List<String>>> getCallerRoleAssignmentsWithResponse(Context context) {
```
Comment (JonathanGiles) [inline sdk/synapse/azure-analytics-synapse-accesscontrol/src/main/java/com/azure/analytics/synapse/accesscontrol/AccessControlAsyncClient.java:329]: We do not need methods taking `Context` arguments in the async clients.

File: sdk/synapse/azure-analytics-synapse-accesscontrol/src/main/java/com/azure/analytics/synapse/accesscontrol/AccessControlAsyncClient.java [hunk 9]
```diff
@@ -0,0 +1,497 @@
...
+     * 
+     * @param nextLink null
+     * @param context The context to associate with this operation.
+     * @throws IllegalArgumentException thrown if parameters fail the validation.
+     * @throws ErrorContractException thrown if the request is rejected by server.
+     * @throws RuntimeException all other wrapped checked exceptions if the request fails to be sent.
+     * @return a list of Synapse roles available.
+     */
+    @ServiceMethod(returns = ReturnType.COLLECTION)
+    public Mono<PagedResponse<SynapseRole>> getRoleDefinitionsNextSinglePage(String nextLink, Context context) {
```
Comment (JonathanGiles) [inline sdk/synapse/azure-analytics-synapse-accesscontrol/src/main/java/com/azure/analytics/synapse/accesscontrol/AccessControlAsyncClient.java:494]: This doesn't follow guidelines on how to provide paginated responses to the user.

File: sdk/synapse/azure-analytics-synapse-accesscontrol/src/main/java/com/azure/analytics/synapse/accesscontrol/AccessControlClient.java [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+     * @param roleId Synapse Built-In Role Id.
+     * @param principalId Object ID of the AAD principal or security-group.
+     * @param continuationToken Continuation token.
+     * @throws IllegalArgumentException thrown if parameters fail the validation.
+     * @throws ErrorContractException thrown if the request is rejected by server.
+     * @throws RuntimeException all other wrapped checked exceptions if the request fails to be sent.
+     * @return a list of role assignments.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public List<RoleAssignmentDetails> getRoleAssignments(String roleId, String principalId, String continuationToken) {
```
Comment (JonathanGiles) [inline sdk/synapse/azure-analytics-synapse-accesscontrol/src/main/java/com/azure/analytics/synapse/accesscontrol/AccessControlClient.java:96]: What is the continuation token here? How is it used?

File: sdk/synapse/azure-analytics-synapse-accesscontrol/src/main/java/com/azure/analytics/synapse/accesscontrol/AccessControlClient.java [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+     * 
+     * @param roleAssignmentId The ID of the role assignment.
+     * @param context The context to associate with this operation.
+     * @throws IllegalArgumentException thrown if parameters fail the validation.
+     * @throws ErrorContractException thrown if the request is rejected by server.
+     * @throws RuntimeException all other wrapped checked exceptions if the request fails to be sent.
+     * @return role assignment by role assignment Id.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public RoleAssignmentDetails getRoleAssignmentById(String roleAssignmentId, Context context) {
```
Comment (JonathanGiles) [inline sdk/synapse/azure-analytics-synapse-accesscontrol/src/main/java/com/azure/analytics/synapse/accesscontrol/AccessControlClient.java:154]: Methods that accept Context in the sync APIs should follow the 'withResponse' pattern and not unwrap the Response from the return type. In this case, I would expect a Response<RoleAssignmentDetails> response type.

File: sdk/synapse/azure-analytics-synapse-accesscontrol/src/main/java/com/azure/analytics/synapse/accesscontrol/AccessControlAsyncClient.java [hunk 12]
```diff
@@ -0,0 +1,497 @@
...
+     * Create role assignment.
+     * 
+     * @param options Role Assignment request details.
+     * @throws IllegalArgumentException thrown if parameters fail the validation.
+     * @throws ErrorContractException thrown if the request is rejected by server.
+     * @throws RuntimeException all other wrapped checked exceptions if the request fails to be sent.
+     * @return role Assignment response details.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<SimpleResponse<RoleAssignmentDetails>> createRoleAssignmentWithResponse(RoleAssignmentOptions options) {
```
Comment (JonathanGiles) [inline sdk/synapse/azure-analytics-synapse-accesscontrol/src/main/java/com/azure/analytics/synapse/accesscontrol/AccessControlAsyncClient.java:65]: RoleAssignmentDetails is a class in the implementation package, which means it cannot be exposed through public API. You need to move RoleAssignmentDetails out of the implementation package if you intend for it to be used by developers.

File: sdk/synapse/azure-analytics-synapse-accesscontrol/src/main/java/com/azure/analytics/synapse/accesscontrol/AccessControlAsyncClient.java (hunk 8 above)
Comment (idear1203) [reply sdk/synapse/azure-analytics-synapse-accesscontrol/src/main/java/com/azure/analytics/synapse/accesscontrol/AccessControlAsyncClient.java:329]: Thanks. Will fix this.

File: sdk/synapse/azure-analytics-synapse-accesscontrol/src/main/java/com/azure/analytics/synapse/accesscontrol/AccessControlAsyncClient.java (hunk 9 above)
Comment (idear1203) [reply sdk/synapse/azure-analytics-synapse-accesscontrol/src/main/java/com/azure/analytics/synapse/accesscontrol/AccessControlAsyncClient.java:494]: Could you please explain a little bit more? What should it look like?

File: sdk/synapse/azure-analytics-synapse-accesscontrol/src/main/java/com/azure/analytics/synapse/accesscontrol/AccessControlClient.java (hunk 11 above)
Comment (idear1203) [reply sdk/synapse/azure-analytics-synapse-accesscontrol/src/main/java/com/azure/analytics/synapse/accesscontrol/AccessControlClient.java:154]: Removed Context

File: sdk/synapse/azure-analytics-synapse-accesscontrol/src/main/java/com/azure/analytics/synapse/accesscontrol/AccessControlAsyncClient.java (hunk 12 above)
Comment (idear1203) [reply sdk/synapse/azure-analytics-synapse-accesscontrol/src/main/java/com/azure/analytics/synapse/accesscontrol/AccessControlAsyncClient.java:65]: Can we expose all generated classes?

File: sdk/synapse/azure-analytics-synapse-accesscontrol/src/main/java/com/azure/analytics/synapse/accesscontrol/AccessControlClient.java (hunk 10 above)
Comment (idear1203) [reply sdk/synapse/azure-analytics-synapse-accesscontrol/src/main/java/com/azure/analytics/synapse/accesscontrol/AccessControlClient.java:96]: It indicates whether there are more results.

**Comment by mitchdenny** [comment]
This PR also needs to have entries added to the ```eng/versioning/version_client.txt``` file so that nightly packages can be produced. Noticed this issue in a failing nightly build.

**Comment by mitchdenny** [comment]
What is the status of this PR? We've got build failures on master because we setup the pipeline for this incoming PR (its trying to build code that doesn't yet exist on master). I'm trying to figure out whether this PR will be landing soon or whether we should look at mothballing the pipeline for now.

**Comment by idear1203** [comment]
@mitchdenny Sorry I have been working on some other tasks recently. Can I close this PR and reopen it sometime later?

**Comment by mitchdenny** [comment]
You can leave the PR. I'll just disable the schedule on the pipeline so it isn't creating noise.

File: sdk/synapse/azure-analytics-synapse-accesscontrol/src/main/java/com/azure/analytics/synapse/accesscontrol/AccessControlAsyncClient.java (hunk 7 above)
Comment (idear1203) [reply sdk/synapse/azure-analytics-synapse-accesscontrol/src/main/java/com/azure/analytics/synapse/accesscontrol/AccessControlAsyncClient.java]: Now the method signature is as below
```java
    public Mono<Response<RoleAssignmentDetails>> getRoleAssignmentByIdWithResponse(String roleAssignmentId) 
```

File: sdk/synapse/azure-analytics-synapse-accesscontrol/src/main/java/com/azure/analytics/synapse/accesscontrol/AccessControlAsyncClient.java (hunk 8 above)
Comment (idear1203) [reply sdk/synapse/azure-analytics-synapse-accesscontrol/src/main/java/com/azure/analytics/synapse/accesscontrol/AccessControlAsyncClient.java]: Fixed

**Comment by idear1203** [comment]
> This PR also needs to have entries added to the `eng/versioning/version_client.txt` file so that nightly packages can be produced. Noticed this issue in a failing nightly build.

Fixed

File: sdk/synapse/azure-analytics-synapse-accesscontrol/src/main/java/com/azure/analytics/synapse/accesscontrol/AccessControlClient.java [hunk 13]
```diff
@@ -0,0 +1,134 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (samvaity) [inline sdk/synapse/azure-analytics-synapse-accesscontrol/src/main/java/com/azure/analytics/synapse/accesscontrol/AccessControlClient.java:1]: All the API's need to add support [context propagation](https://azure.github.io/azure-sdk/java_design.html#common-service-client-patterns).

File: sdk/synapse/azure-analytics-synapse-accesscontrol/src/main/java/com/azure/analytics/synapse/accesscontrol/AccessControlClient.java [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+
+    /**
+     * List role assignments.
+     *
+     * @throws ErrorContractException thrown if the request is rejected by server.
+     * @throws RuntimeException all other wrapped checked exceptions if the request fails to be sent.
+     * @return a list of role assignments.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public List<RoleAssignmentDetails> getRoleAssignments() {
```
Comment (samvaity) [inline sdk/synapse/azure-analytics-synapse-accesscontrol/src/main/java/com/azure/analytics/synapse/accesscontrol/AccessControlClient.java:66]: Should the Api's returning a  `List` be updated to use prefix `list..` according to [this](https://azure.github.io/azure-sdk/java_design.html#common-service-client-patterns)?

File: sdk/synapse/azure-analytics-synapse-accesscontrol/src/main/java/com/azure/analytics/synapse/accesscontrol/models/AccessControlsGetRoleAssignmentsHeaders.java [hunk 15]
```diff
@@ -0,0 +1,38 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+// Code generated by Microsoft (R) AutoRest Code Generator.
```
Comment (samvaity) [inline sdk/synapse/azure-analytics-synapse-accesscontrol/src/main/java/com/azure/analytics/synapse/accesscontrol/models/AccessControlsGetRoleAssignmentsHeaders.java:3]: The generated models should belong to `.implementation` package?

Comment (idear1203) [reply sdk/synapse/azure-analytics-synapse-accesscontrol/src/main/java/com/azure/analytics/synapse/accesscontrol/models/AccessControlsGetRoleAssignmentsHeaders.java:3]: We want to expose them to developers. See words of @JonathanGiles at below
> RoleAssignmentDetails is a class in the implementation package, which means it cannot be exposed through public API. You need to move RoleAssignmentDetails out of the implementation package if you intend for it to be used by developers.

File: sdk/synapse/azure-analytics-synapse-accesscontrol/src/main/java/com/azure/analytics/synapse/accesscontrol/AccessControlClient.java (hunk 13 above)
Comment (idear1203) [reply sdk/synapse/azure-analytics-synapse-accesscontrol/src/main/java/com/azure/analytics/synapse/accesscontrol/AccessControlClient.java:1]: Fixed

**Comment by mitchdenny** [comment]
I don't think you meant to include the ```jacoco.exec``` file in this PR?

**Comment by idear1203** [comment]
> I don't think you meant to include the `jacoco.exec` file in this PR?

Good catch! Will remove in the next push

File: eng/code-quality-reports/src/main/resources/checkstyle/checkstyle-suppressions.xml [hunk 16]
```diff
@@ -298,6 +298,10 @@
   <suppress checks="com.azure.tools.checkstyle.checks.ServiceClientCheck"
             files="com.azure.messaging.eventhubs.EventHubAsyncClient.java"/>
 
+  <!-- TODO: Synapse access control API does not support PagedResponse and Response. -->
```
Comment (conniey) [inline eng/code-quality-reports/src/main/resources/checkstyle/checkstyle-suppressions.xml:301]: Can you create a bug for this and link it here so we don't lose track?

**Comment by jianghaolu** [comment]
Migrating to https://github.com/Azure/azure-sdk-for-java/pull/17909 and regenerated with a few latest autorest updates.

**Pull request closed without merging**